### PR TITLE
Add certificate based authentication in check_sync and pldap/pldaps support

### PR DESCRIPTION
**Pull request by damiencassu**
Hello,

I would like to merge the following changes aimed to:

- In the check_sync(), if sasl/external mechanism is detected in the olcsyncrepl statement, try certificate based authentication to retrieve the sync state
- Add the support of proxyfied ldap endpoints (pldap:// and pldaps://) in case you want to put your slapd instance behind a load balancer and still preserve the client IP using the proxy protocol v2

I merged my local repository with the lastest version of yours, so normaly the changes validated in the PR#56 should be already there :)

Thanks !

### CODE REVIEW

**Comment by damiencassu** [comment]
Hello @coudot, 

By any chance, would it be possible to have your feedback/review on this PR ?

Thanks a lot !

**Comment by davidcoutadeur** [comment]
Hello @damiencassu,

I have done a review some time ago, and left a few remarks in the code.

Please give a look at https://github.com/ltb-project/slapd-cli/pull/59/files/08631d2a7f31d866bb216ff678b5f77964ba00de

Globally, the PR is not generic enough to be integrated as it is.

**Comment by damiencassu** [comment]
Hello @davidcoutadeur , 

Thanks for your message !

I'm sorry I didn't get any notification about a review or code comments...and when I look at your link I only see a diff between the changed files, I cannot see any comments on improvements to do on a specific line/part ... am I missing something ?

Rgs,

**Comment by davidcoutadeur** [comment]
> Hello @davidcoutadeur ,
> 
> Thanks for your message !
> 
> I'm sorry I didn't get any notification about a review or code comments...and when I look at your link I only see a diff between the changed files, I cannot see any comments on improvements to do on a specific line/part ... am I missing something ?
> 
> Rgs,

I am sorry, it's my fault: I didn't applied the comments, so they were only visible on my side...
I apologize for the delay!

File: slapd-cli [hunk 1]
```diff
@@ -415,12 +420,12 @@ start_slapd() {
 	# It's time to start slapd
 
 	if [ -n "$1" ]; then
-		$SLAPD_BIN -h "$SLAPD_SERVICES" $SLAPD_PARAMS -d $1
+		$SLAPD_BIN -h "$SLAPD_SERVICES $SLAPD_PROXY_SERVICES" $SLAPD_PARAMS -d $1
```
[the remark this answers is not in the document]
Comment (damiencassu) [reply slapd-cli:423]: I understand your point.
I can work on a solution to make it not mandatory (using a boolean variable to explicitly ask for pldap and pldaps bind for instance). Will that be ok for you ?

**Comment by damiencassu** [comment]
No problem :)
Thanks for posting them now !

File: slapd-cli (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidcoutadeur) [reply slapd-cli:423]: It won't be easy, but if you find a way, that's ok. The goal is to make pldap:/// and pldaps:/// connections disabled by default, for both existing and new installations.

File: slapd-cli [hunk 2]
```diff
@@ -1365,10 +1377,29 @@ check_sync() {
...
+			syncrepl_get_value "${PROV}" "tls_cacert"
+			PCACERT=$RETVAL
+			syncrepl_get_value "${PROV}" "tls_cert"
+			PCERT=$RETVAL
+			syncrepl_get_value "${PROV}" "tls_key"
+			PKEY=$RETVAL
 			echo "Checking contextCSN for suffix ${SUFFIX} on remote host ${PHOST}"
 			# Get remote provider contextCSN (one contextCSN per line)
+			# If certificate based authentication is used, use it
+			if [ "$PBINDMTD" == "sasl" ] && [ "$PSASLMECH" == "external" ]; then
```
[the remark this answers is not in the document]
Comment (davidcoutadeur) [reply slapd-cli:1395]: Either we are able to guess all possible methods in the given arguments, or we must have a parameter set by the admin. I have no strong opinion on this.

**Comment by damiencassu** [comment]
Hello @davidcoutadeur, 

Please find above two new commits to take your remarks into account.
Let me know what you think about them !

Have a nice day !

**Comment by davidcoutadeur** [comment]
> Please find above two new commits to take your remarks into account. Let me know what you think about them !

Thanks for your reply. I'll give a look at the beginning of next week. (week 5)

**Comment by davidcoutadeur** [comment]
Hello @damiencassu,

I have reviewed your PR. Seems better indeed.
I have created a dedicated branch here: https://github.com/ltb-project/slapd-cli/commits/pldap-and-certificate-auth/ where I have pulled your commits, and added 4 additional commits:
* f1af6dc simplify bind by certificate (and remove $BIND_METHOD)
* 047fc65 simplify LOCAL_URL_SSL computation
* e50e087 fix extra spaces and tabs
* 113b81a remove PLDAP_ENABLED switch and provide a commented example of pldap url

The main changes are:
* removal of `PLDAP_ENABLED` switch. I don't think we need this. The pldap and pldaps url can simply be declared in SLAPD_SERVICES parameter. I have provided a commented example in `slapd-cli.conf`
* I have improved `LOCAL_URL_SSL` (because ldaps url can be anywhere in the SLAPD_SERVICES string)
* finally, I changed my mind about the `$BIND_METHOD` parameter. It seems overkill because we already need to get key and cert variables from configuration. For now, I assume that if we have a tls_cert or tls_key defined in the configuration we can do a client cert authentication by default.

I don't have an openldap instance with client cert authentication. Could you review and test this new version?

Last point: for next time you create a PR, it would be better to separate features in different PR (for example, one for pldap/pldaps support, and another for client cert authentication)

**Comment by davidcoutadeur** [comment]
Final PR in #65

**Pull request closed without merging**